<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Public · JWAS.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.png" alt="JWAS.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit">JWAS.jl</span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><a class="tocitem" href="../../theory/theory/">Some Theory</a></li><li><a class="tocitem" href="../../citing/citing/">Citing</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../getstarted/">Get Started</a></li><li><a class="tocitem" href="../workflow/">Workflow</a></li><li class="is-active"><a class="tocitem" href>Public</a><ul class="internal"><li><a class="tocitem" href="#Index-1"><span>Index</span></a></li><li><a class="tocitem" href="#Public-Interface-1"><span>Public Interface</span></a></li></ul></li><li><a class="tocitem" href="../internals/">Internals</a></li></ul></li><li><span class="tocitem">Examples</span><ul><li><a class="tocitem" href="../../examples/examples/">Examples</a></li></ul></li><li><a class="tocitem" href="../../FrequentlyAskedQuestions/FrequentlyAskedQuestions/">Frequently Asked Questions</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Manual</a></li><li class="is-active"><a href>Public</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Public</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/reworkhow/JWAS.jl/blob/master/docs/src/manual/public.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Public-functions-1"><a class="docs-heading-anchor" href="#Public-functions-1">Public functions</a><a class="docs-heading-anchor-permalink" href="#Public-functions-1" title="Permalink"></a></h1><p>Documentation for JWAS.jl&#39;s public interface. Below are functions available to general users.</p><h2 id="Index-1"><a class="docs-heading-anchor" href="#Index-1">Index</a><a class="docs-heading-anchor-permalink" href="#Index-1" title="Permalink"></a></h2><ul><li><a href="#JWAS.GWAS"><code>JWAS.GWAS</code></a></li><li><a href="#JWAS.add_genotypes"><code>JWAS.add_genotypes</code></a></li><li><a href="#JWAS.build_model"><code>JWAS.build_model</code></a></li><li><a href="#JWAS.outputEBV"><code>JWAS.outputEBV</code></a></li><li><a href="#JWAS.outputMCMCsamples"><code>JWAS.outputMCMCsamples</code></a></li><li><a href="#JWAS.runMCMC"><code>JWAS.runMCMC</code></a></li><li><a href="#JWAS.set_covariate"><code>JWAS.set_covariate</code></a></li><li><a href="#JWAS.set_random"><code>JWAS.set_random</code></a></li><li><a href="#JWAS.showMME"><code>JWAS.showMME</code></a></li><li><a href="#JWAS.solve"><code>JWAS.solve</code></a></li></ul><h2 id="Public-Interface-1"><a class="docs-heading-anchor" href="#Public-Interface-1">Public Interface</a><a class="docs-heading-anchor-permalink" href="#Public-Interface-1" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="JWAS.build_model" href="#JWAS.build_model"><code>JWAS.build_model</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">build_model(model_equations::AbstractString,R=false; df::AbstractFloat=4.0)</code></pre><ul><li>Build a model from <strong>model equations</strong> with the residual variance <strong>R</strong>. In Bayesian analysis, <strong>R</strong> is the mean for the prior assigned for the residual variance with degree of freedom <strong>df</strong>, defaulting to 4.0. If <strong>R</strong> is not provided, a value is calculated from responses (phenotypes).</li><li>By default, all variabels in model<em>equations are factors (categorical) and fixed. Set variables to be covariates (continuous) or random using functions `set</em>covariate()<code>or</code>set_random()`.</li></ul><pre><code class="language-julia">#single-trait
model_equations = &quot;BW = intercept + age + sex&quot;
R               = 6.72
models          = build_model(model_equations,R);

#multi-trait
model_equations = &quot;BW = intercept + age + sex
                   CW = intercept + litter&quot;;
R               = [6.72   24.84
                   24.84  708.41]
models          = build_model(model_equations,R);</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9f7f9fe897a080af3f1137112720bcc88d9322f9/src/1.JWAS/src/build_MME.jl#L14-L36">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="JWAS.set_covariate" href="#JWAS.set_covariate"><code>JWAS.set_covariate</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">set_covariate(model::MME,variables::AbstractString...)</code></pre><ul><li>set <strong>variables</strong> as covariates; <strong>model</strong> is the output of function <code>build_model()</code>.</li></ul><pre><code class="language-julia">#After running build_model, variabels age and year can be set to be covariates as
set_covariate(model,&quot;age&quot;,&quot;year&quot;)
#or
set_covariate(model,&quot;age year&quot;)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9f7f9fe897a080af3f1137112720bcc88d9322f9/src/1.JWAS/src/build_MME.jl#L123-L134">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="JWAS.set_random" href="#JWAS.set_random"><code>JWAS.set_random</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">set_random(mme::MME,randomStr::AbstractString,ped::Pedigree, G;df=4)</code></pre><ul><li>set variables as random polygenic effects with pedigree information <strong>ped</strong>. and variances <strong>G</strong>.</li><li><strong>G</strong> is the mean for the prior assigned for the variance with degree of freedom <strong>df</strong>, defaulting to 4.0. If <strong>G</strong> is not provided, a value is calculated from responses (phenotypes).</li></ul><pre><code class="language-julia">#single-trait (example 1)
model_equation  = &quot;y = intercept + age + animal&quot;
model           = build_model(model_equation,R)
ped             = get_pedigree(pedfile)
G               = 1.6
set_random(model,&quot;animal&quot;, ped, G)

#single-trait (example 2)
model_equation  = &quot;y = intercept + age + animal + animal*age&quot;
model           = build_model(model_equation,R)
ped             = get_pedigree(pedfile)
G               = [1.6   0.2
                   0.2  1.0]
set_random(model,&quot;animal animal*age&quot;, ped,G)

#multi-trait
model_equations = &quot;BW = intercept + age + sex + animal
                   CW = intercept + age + sex + animal&quot;
model           = build_model(model_equations,R);
ped             = get_pedigree(pedfile);
G               = [6.72   2.84
                   2.84  8.41]
set_random(model,&quot;animal&quot;,ped,G)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9f7f9fe897a080af3f1137112720bcc88d9322f9/src/1.JWAS/src/random_effects.jl#L5-L39">source</a></section><section><div><pre><code class="language-julia">set_random(mme::MME,randomStr::AbstractString,G;Vinv=0,names=[],df=4)</code></pre><ul><li>set variables as random effects, defaulting to i.i.d effects, with variances <strong>G</strong>.</li><li><strong>G</strong> is the mean for the prior assigned for the variance with degree of freedom <strong>df</strong>, defaulting to 4.0. If <strong>G</strong> is not provided, a value is calculated from responses (phenotypes).</li><li>the random effects are assumed to be i.i.d by default and it can be defined with any (inverse of) covariance structure <strong>Vinv</strong> with its index (row names) provided by <strong>names</strong>.</li></ul><pre><code class="language-julia">#single-trait (i.i.d randome effects)
model_equation  = &quot;y = intercept + litter + sex&quot;
model           = build_model(model_equation,R)
G               = 0.6
set_random(model,&quot;litter&quot;,G)

#multi-trait (i.i.d randome effects)
model_equations = &quot;BW = intercept + litter + sex
                   CW = intercept + litter + sex&quot;
model           = build_model(model_equations,R);
G               = [3.72  1.84
                   1.84  3.41]
set_random(model,&quot;litter&quot;,G)

#single-trait (randome effects with specific covariance structures)
model_equation  = &quot;y = intercept + litter + sex&quot;
model           = build_model(model_equation,R)
V               = [1.0  0.5 0.25
                   0.5  1.0 0.5
                   0.25 0.5 1.0]
G               = 0.6
set_random(model,&quot;litter&quot;,G,Vinv=inv(V),names=[a1;a2;a3])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9f7f9fe897a080af3f1137112720bcc88d9322f9/src/1.JWAS/src/random_effects.jl#L52-L85">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="JWAS.PedModule.get_pedigree" href="#JWAS.PedModule.get_pedigree"><code>JWAS.PedModule.get_pedigree</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">get_pedigree(pedfile::AbstractString;header=false,separator=&#39;,&#39;)</code></pre><ul><li>Get pedigree informtion from a pedigree file with <strong>header</strong> (defaulting to <code>false</code>) and <strong>separator</strong> (defaulting to <code>,</code>).</li><li>Pedigree file format:</li></ul><pre><code class="language-none">a,0,0
c,a,b
d,a,c</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9f7f9fe897a080af3f1137112720bcc88d9322f9/src/2.PedModule/src/PedModule.jl#L10-L21">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="JWAS.add_genotypes" href="#JWAS.add_genotypes"><code>JWAS.add_genotypes</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">add_genotypes(mme::MME,M::Union{AbstractString,Array{Float64,2},Array{Float32,2},Array{Any,2},DataFrames.DataFrame},G=false;
              header=false,rowID=false,separator=&#39;,&#39;,
              center=true,G_is_marker_variance=false,df=4)</code></pre><ul><li>Get marker informtion from a genotype file or an nxp Matrix M of genotypes (Array or DataFrame), where n is the number of individuals and p is the number of markers. This file/matrix needs to be column-wise sorted by marker positions.</li><li><strong>G</strong> is the mean for the prior assigned for the genomic variance with degree of freedom <strong>df</strong>, defaulting to 4.0. If <strong>G</strong> is not provided, a value is calculated from responses (phenotypes)</li><li>If a text file is provided, the file format should be:<ul><li><pre><code class="language-none">Animal,marker1,marker2,marker3,marker4,marker5
S1,1,0,1,1,1
D1,2,0,2,2,1
O1,1,2,0,1,0
O3,0,0,2,1,1</code></pre></li></ul></li><li>If an nxp Matrix of genotypes (Array or DataFrame) is provided, where n is the number of individuals and p is the number of markers,<ul><li>This matrix needs to be column-wise sorted by marker positions.</li><li>rowID is a vector of individual IDs, e.g.,rowID=[&quot;a1&quot;,&quot;b2&quot;,&quot;c1&quot;]; if it is omitted, IDs will be set to 1:n</li><li>header is a header vector such as [&quot;id&quot;; &quot;mrk1&quot;; &quot;mrk2&quot;;...;&quot;mrkp&quot;]. If omitted, marker names will be set to 1:p</li></ul></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9f7f9fe897a080af3f1137112720bcc88d9322f9/src/1.JWAS/src/markers/readgenotypes.jl#L1-L21">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="JWAS.runMCMC" href="#JWAS.runMCMC"><code>JWAS.runMCMC</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">runMCMC(model::MME,df::DataFrame;
        weight                   = false,
        ### MCMC
        chain_length             = 1000,
        starting_value           = false,
        burnin                   = 0,
        output_samples_frequency = chain_length/1000,
        output_folder            = results,
        update_priors_frequency  = 0,
        ### Methods
        estimate_variance        = true,
        estimateScale            = false,
        single_step_analysis     = false,
        pedigree                 = false,
        categorical_trait        = false,
        missing_phenotypes       = true,
        constraint               = false,
        causal_structure         = false,
        ### Genomic Prediction
        outputEBV                = true,
        output_heritability      = true,
        ### MISC
        seed                     = false,
        printout_model_info      = true,
        printout_frequency       = 0)</code></pre><p><strong>Run MCMC for Bayesian Linear Mixed Models with or without estimation of variance components.</strong></p><ul><li>Markov chain Monte Carlo<ul><li>The first <code>burnin</code> iterations are discarded at the beginning of a MCMC chain of length <code>chain_length</code>.</li><li>Save MCMC samples every <code>output_samples_frequency</code> iterations, defaulting to <code>chain_length/1000</code>, to a folder <code>output_folder</code>, defaulting to <code>results</code>. MCMC samples for hyperparametes (variance componets) and marker effects are saved by default. MCMC samples for location parametes can be saved using <code>output_MCMC_samples()</code>. Note that saving MCMC samples too frequently slows down the computation.</li><li>The <code>starting_value</code> can be provided as a vector for all location parameteres and marker effects, defaulting to <code>0.0</code>s. The order of starting values for location parameters and marker effects should be the order of location parameters in the Mixed Model Equation for all traits (This can be obtained by getNames(model)) and then markers for all traits (all markers for trait 1 then all markers for trait 2...).</li><li>Miscellaneous Options<ul><li>Priors are updated every <code>update_priors_frequency</code> iterations, defaulting to <code>0</code>.</li></ul></li></ul></li><li>Methods<ul><li>Single step analysis is allowed if <code>single_step_analysis</code> = <code>true</code> and <code>pedigree</code> is provided.</li><li>Variance components are estimated if <code>estimate_variance</code>=true, defaulting to <code>true</code>.</li><li>Miscellaneous Options<ul><li>Missing phenotypes are allowed in multi-trait analysis with <code>missing_phenotypes</code>=true, defaulting to <code>true</code>.</li><li>Catogorical Traits are allowed if <code>categorical_trait</code>=true, defaulting to <code>false</code>. Phenotypes should be coded as 1,2,3...</li><li>If <code>constraint</code>=true, defaulting to <code>false</code>, constrain residual covariances between traits to be zeros.</li><li>If <code>causal_structure</code> is provided, e.g., causal_structure = [0.0,0.0,0.0;1.0,0.0,0.0;1.0,0.0,0.0] for trait 2 -&gt; trait 1 and trait 3 -&gt; trait 1, phenotypic causal networks will be incorporated using structure equation models.</li></ul></li></ul></li><li>Genomic Prediction<ul><li>Individual estimted breeding values (EBVs) and prediction error variances (PEVs) are returned if <code>outputEBV</code>=true, defaulting to <code>true</code>. Heritability and genetic</li></ul>variances are returned if <code>output_heritability</code>=<code>true</code>, defaulting to <code>true</code>. Note that estimation of heritability is computaionally intensive.</li><li>Miscellaneous Options<ul><li>Print out the model information in REPL if <code>printout_model_info</code>=true; print out the monte carlo mean in REPL with <code>printout_frequency</code>, defaulting to <code>false</code>.</li><li>If <code>seed</code>, defaulting to <code>false</code>, is provided, a reproducible sequence of numbers will be generated for random number generation.</li></ul></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9f7f9fe897a080af3f1137112720bcc88d9322f9/src/1.JWAS/src/JWAS.jl#L64-L121">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="JWAS.outputEBV" href="#JWAS.outputEBV"><code>JWAS.outputEBV</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">outputEBV(model,IDs::Array)</code></pre><p>Output estimated breeding values and prediction error variances for IDs.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9f7f9fe897a080af3f1137112720bcc88d9322f9/src/1.JWAS/src/output.jl#L11-L15">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="JWAS.outputMCMCsamples" href="#JWAS.outputMCMCsamples"><code>JWAS.outputMCMCsamples</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">outputMCMCsamples(mme::MME,trmStrs::AbstractString...)</code></pre><p>Get MCMC samples for specific location parameters.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9f7f9fe897a080af3f1137112720bcc88d9322f9/src/1.JWAS/src/output.jl#L21-L25">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="JWAS.showMME" href="#JWAS.showMME"><code>JWAS.showMME</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">showMME(mme::MME,df::DataFrame)</code></pre><ul><li>Show left-hand side and right-hand side of mixed model equations (no markers).</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9f7f9fe897a080af3f1137112720bcc88d9322f9/src/1.JWAS/src/build_MME.jl#L337-L341">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="JWAS.solve" href="#JWAS.solve"><code>JWAS.solve</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">solve(mme::MME,df::DataFrame;solver=&quot;default&quot;,printout_frequency=100,tolerance = 0.000001,maxiter = 5000)</code></pre><ul><li>Solve the mixed model equations (no marker information) without estimating variance components.</li></ul><p>Available solvers include <code>default</code>, <code>Jacobi</code>, <code>Gauss-Seidel</code>, and <code>Gibbs sampler</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9f7f9fe897a080af3f1137112720bcc88d9322f9/src/1.JWAS/src/iterative_solver/solver.jl#L1-L6">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="JWAS.GWAS" href="#JWAS.GWAS"><code>JWAS.GWAS</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">GWAS(marker_effects_file,model;header=true,window_size=100,threshold=0.001)</code></pre><p>run genomic window-based GWAS without a map file</p><ul><li>MCMC samples of marker effects are stored in <strong>marker<em>effects</em>file</strong> with delimiter &#39;,&#39;.</li><li><strong>window_size</strong> is either a constant (identical number of markers in each window) or an array of number of markers in each window</li><li><strong>model</strong> is either the model::MME used in analysis or the genotypic covariate matrix M::Array</li><li>File format:</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9f7f9fe897a080af3f1137112720bcc88d9322f9/src/3.GWAS/src/GWAS_deprecated.jl#L1-L11">source</a></section><section><div><pre><code class="language-julia">GWAS(marker_effects_file;header=true)</code></pre><p>Compute the model frequency for each marker (the probability the marker is included in the model) using samples of marker effects stored in <strong>marker<em>effects</em>file</strong>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9f7f9fe897a080af3f1137112720bcc88d9322f9/src/3.GWAS/src/GWAS.jl#L3-L7">source</a></section><section><div><pre><code class="language-julia">GWAS(model,map_file,marker_effects_file...;
     window_size = &quot;1 Mb&quot;,sliding_window = false,
     GWAS = true, threshold = 0.001,
     genetic_correlation = false,
     header = true)</code></pre><p>run genomic window-based GWAS</p><ul><li>MCMC samples of marker effects are stored in <strong>marker<em>effects</em>file</strong> with delimiter &#39;,&#39;.</li><li><strong>model</strong> is either the model::MME used in analysis or the genotype cavariate matrix M::Array</li><li><strong>map_file</strong> has the (sorted) marker position information with delimiter &#39;,&#39;. If the map file is not provided, a fake map file will be generated with 100 markers in each 1 Mb window.</li><li>If two <strong>marker<em>effects</em>file</strong> are provided, and <strong>genetic_correlation</strong> = true, genomic correlation for each window is calculated.</li><li>Statistics are computed for nonoverlapping windows of size <em>window_size</em> by default. If <strong>sliding_window</strong> = true, those for overlapping sliding windows are calculated.</li><li>map file format:</li></ul><pre><code class="language-none">markerID,chromosome,position
m1,1,16977
m2,1,434311
m3,1,1025513
m4,2,70350
m5,2,101135</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9f7f9fe897a080af3f1137112720bcc88d9322f9/src/3.GWAS/src/GWAS.jl#L33-L60">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../workflow/">« Workflow</a><a class="docs-footer-nextpage" href="../internals/">Internals »</a></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Wednesday 14 October 2020 04:54">Wednesday 14 October 2020</span>. Using Julia version 1.1.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
